File: src/entities/CollectionItem.ts
import {
  Column,
  Entity,
  Index,
  JoinColumn,
  ManyToMany,
  ManyToOne,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { Collection } from './Collection';
import { User } from './User';
import { Platforms } from './Platforms';
import { EntityModel } from './EntityModel';

@Entity()
export class CollectionItem {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Index()
  @Column({ nullable: false })
  name: string;

  @Index()
  @Column('simple-array', { array: true, nullable: true })
  images: string[];

  @Index()
  @Column({ nullable: true })
  purchase_date: Date;

  @Index()
  @Column({ nullable: false, default: 0 })
  purchase_price: number;

  @Index()
  @Column({ nullable: false, default: false })
  sealed: boolean;

  @Index()
  @Column({ nullable: true })
  edition: string;

  @Index()
  @Column('simple-array', { array: true, nullable: true })
  copy_number: string[];

  @Index()
  @Column({ default: false, select: false })
  deleted: boolean;

  @ManyToMany(() => Collection, (collection) => collection.collectionItems, {
    onDelete: 'CASCADE',
  })
  collections: Collection[];

  @Index()
  @ManyToOne(() => User, (user) => user.login, {
    onDelete: 'CASCADE',
  })
  @JoinColumn({ name: 'owner' })
  owner: string;

  @ManyToOne(() => Platforms, (platform) => platform.id, {
    onDelete: 'CASCADE',
  })
  @JoinColumn({ name: 'platform' })
  platform: string;

  @Index()
  @Column({
    nullable: true,
  })
  description: string;

  @Index()
  @Column({
    nullable: true,
  })
  shipping_cost: number;

  @Index()
  @Column({
    nullable: true,
  })
  transliteration: string;

  @ManyToMany(() => EntityModel, (entity) => entity.collection_item)
  entities: EntityModel[];
}

File: internal/models/collectionItem.go
package models

import (
	"github.com/google/uuid"
	"github.com/lib/pq"
	"gorm.io/gorm"
)

type EntityModelCollectionItemCollectionItem struct {
	EntityModelId    uuid.UUID `gorm:"primaryKey;column:entityModelId"`
	CollectionItemId uuid.UUID `gorm:"primaryKey;column:collectionItemId"`
}

type CollectionItem struct {
	gorm.Model
	Id              uuid.UUID `gorm:"type:uuid;primaryKey;default:uuid_generate_v4()"`
	Name            string
	Description     string
	Images          pq.StringArray `gorm:"type:text[]"`
	PurchaseDate    string
	PurchasePrice   int
	Sealed          bool
	Edition         string
	CopyNumber      pq.StringArray `gorm:"type:text[]"`
	Deleted         bool
	ShippingCost    int
	Transliteration string
	Collections     []Collection  `gorm:"many2many:collection_collection_items;constraint:OnDelete:CASCADE;"`
	Owner           User          `gorm:"constraint:OnDelete:CASCADE;"`
	Platform        Platforms     `gorm:"constraint:OnDelete:CASCADE;"`
	Entities        []EntityModel `gorm:"many2many:entity_model_collection_item_collection_item;constraint:OnDelete:CASCADE;"`
}

func (CollectionItem) TableName() string {
	return "collection_item"
}

func (EntityModelCollectionItemCollectionItem) TableName() string {
	return "entity_model_collection_item_collection_item"
}
